src/gar.lua

```diff
@@ -937,48 +1002,135 @@ function gar.instantDelegateWithdrawal(from, gatewayAddress, vaultId, currentTim
+	local gateway = gar.getGateway(gatewayAddress)
+	if gateway == nil then
+		error("Gateway not found")
 	end
 
-	-- Process the instant withdrawal
-	local expeditedWithdrawalFee, amountToWithdraw, penaltyRate =
-		processInstantWithdrawal(vault.balance, elapsedTime, totalWithdrawalTime, from)
+	-- Only allow modification of the allow list when allowDelegatedStaking is set to false or a current allow list is in place
+	if gateway.settings.allowDelegatedStaking == true or not gateway.settings.allowedDelegatesLookup then
+		error("Allow listing only possible when allowDelegatedStaking is set to 'allowlist'")
+	end
 
-	-- Remove the vault after withdrawal
-	gateway.vaults[vaultId] = nil
+	assert(gateway.settings.allowedDelegatesLookup, "allowedDelegatesLookup should not be nil")
 
-	-- Update the gateway
-	GatewayRegistry[from] = gateway
+	local removedDelegates = {}
+	for _, delegateToDisallow in ipairs(delegates) do
+		if not utils.isValidAOAddress(delegateToDisallow) then
+			error("Invalid delegate address: " .. delegateToDisallow)
+		end
+
+		-- Skip over delegates that are not in the allow list
+		if gateway.settings.allowedDelegatesLookup[delegateToDisallow] then
+			gateway.settings.allowedDelegatesLookup[delegateToDisallow] = nil
+			table.insert(removedDelegates, delegateToDisallow)
+		end
+		-- Kick the delegate off the gateway if necessary
+		local ban = true
+		gar.kickDelegateFromGateway(delegateToDisallow, gateway, msgId, currentTimestamp, ban)
+	end
+
+	GatewayRegistry[gatewayAddress] = gateway
 	return {
-		gateway = gar.getGateway(from),
-		elapsedTime = elapsedTime,
-		remainingTime = totalWithdrawalTime - elapsedTime,
-		penaltyRate = penaltyRate,
-		expeditedWithdrawalFee = expeditedWithdrawalFee,
-		amountWithdrawn = amountToWithdraw,
+		gateway = gateway,
+		removedDelegates = removedDelegates,
 	}
 end
 
+--- Vaults delegate's tokens and updates delegate and gateway staking balances
+function gar.kickDelegateFromGateway(delegateAddress, gateway, msgId, currentTimestamp, ban)
+	local delegate = gateway.delegates[delegateAddress]
+	if not delegate then
+		return
+	end
+
+	if not delegate.vaults then
+		delegate.vaults = {}
+	end
+
+	if delegate.delegatedStake > 0 then
+		delegate.vaults[msgId] = {
+			balance = delegate.delegatedStake,
+			startTimestamp = currentTimestamp,
+			endTimestamp = currentTimestamp + gar.getSettings().delegates.withdrawLengthMs,
+		}
+		-- reduce gateway stake and set this delegate stake to 0
+		gateway.totalDelegatedStake = gateway.totalDelegatedStake - delegate.delegatedStake
+		delegate.delegatedStake = 0
+	end
+
+	if not ban and gar.delegationAllowlistedOnGateway(gateway) then
+		gateway.settings.allowedDelegatesLookup[delegateAddress] = true
+	end
+end
+
+function gar.delegationAllowlistedOnGateway(gateway)
+	return gateway.settings.allowedDelegatesLookup ~= nil
+end
+
+function gar.delegateAllowedToStake(delegateAddress, gateway)
+	if not gar.delegationAllowlistedOnGateway(gateway) then
```

Comment:
Noticed this function does not check gateway.settings.allowDelegatedStaking. I see in the code that we check for this outside of this function in places but seems safer to me to add a check here too (if there isn't going to be a later gas cost associated). Otherwise might be good to add a comment that gateway.settings.allowDelegatedStaking should be checked prior to using this function.